lib/Middleware.js

```diff
@@ -0,0 +1,212 @@
+/*!
+ * Copyright (c) 2018 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+// FIXME: use did-io
+const axios = require('axios');
+const didio = {get: () => {}};
+const httpSignatureHeader = require('http-signature-header');
+const httpSignatureCrypto = require('http-signature-crypto');
+const jsonld = require('jsonld');
+const url = require('url');
+const HttpSignatureError = require('./HttpSignatureError');
+const {Strategy} = require('passport-strategy');
+
+module.exports = class Middleware extends Strategy {
+  constructor({name}) {
+    super();
+    this.name = name || 'http-signature-strategy';
+    this._helpers = {};
+    this.dereferenceUrlScheme = {did: true, https: true};
+  }
+
+  async authenticate(req) {
+    const self = this;
+    const user = {};
+
+    // parse the req
+    let parsed;
+    try {
+      parsed = httpSignatureHeader.parseRequest(req);
+    } catch(err) {
+      return self.error(err);
+    }
+
+    const publicKey = {id: parsed.keyId};
+
+    const getKey = self.use('getKey') || self.getPublicKey.bind(self);
+
+    let keyDoc;
+    try {
+      keyDoc = await getKey(publicKey);
+    } catch(err) {
+      return self.error(err);
+    }
+
+    if(keyDoc.revoked && _timestampBeforeNow(keyDoc.revoked)) {
+      return self.error(new HttpSignatureError(
+        'Public key has been revoked.', 'InvalidStateError'));
+    }
+
+    try {
+      await self._verifySignature({keyDoc, parsed});
+    } catch(err) {
+      return self.error(err);
+    }
+
+    // validate the `host` header
+
+    // FIXME: this seems a bit clunky.  Other options range from passing in a
+    // synchronous function to just a string for comparison
+
+    const validateHost = self.use('validateHost');
+    if(!validateHost) {
+      return self.error(new HttpSignatureError(
+        'A `validateHost` function must be defined.', 'InvalidStateError'));
+    }
+    try {
+      await validateHost(req.headers.host);
```

Comment:
Seems to me that `validateHeaders` should validate headers.  Should be do `validateRequest` instead and just pass `req`?

Reply:
`validateRequest` is fine. I guess if we had a client-side lib it would then be `validateResponse`. I was thinking along the lines of unifying for if/when that happens, but `validateRequest` will do.